File: app/src/main/java/com/bopr/android/smailer/util/SharedPreferencesWrapper.java
package com.bopr.android.smailer.util;

import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static android.content.SharedPreferences.Editor;
import static android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import static com.bopr.android.smailer.util.TextUtil.commaJoin;
import static com.bopr.android.smailer.util.TextUtil.commaSplit;
import static com.bopr.android.smailer.util.Util.setOf;
import static java.util.Arrays.asList;

@SuppressWarnings({"UnusedReturnValue", "unused", "WeakerAccess"})
public class SharedPreferencesWrapper {

    private final SharedPreferences wrappedPreferences;

    public SharedPreferencesWrapper(@NonNull SharedPreferences wrappedPreferences) {
        this.wrappedPreferences = wrappedPreferences;
    }

    @NonNull
    public Map<String, ?> getAll() {
        return wrappedPreferences.getAll();
    }

    @Nullable
    public String getString(@NonNull String key) {
        return wrappedPreferences.getString(key, null);
    }

    @NonNull
    public String getString(@NonNull String key, @NonNull String defValue) {
        return wrappedPreferences.getString(key, defValue);
    }

    @NonNull
    public Set<String> getStringSet(@NonNull String key) {
        /* should be a copy of values set.
           see: https://stackoverflow.com/questions/17469583/setstring-in-android-sharedpreferences-does-not-save-on-force-close */
        return new LinkedHashSet<>(wrappedPreferences.getStringSet(key, setOf()));
    }

    public int getInt(@NonNull String key, int defValue) {
        return wrappedPreferences.getInt(key, defValue);
    }

    public long getLong(@NonNull String key, long defValue) {
        return wrappedPreferences.getLong(key, defValue);
    }

    public float getFloat(@NonNull String key, float defValue) {
        return wrappedPreferences.getFloat(key, defValue);
    }

    public boolean getBoolean(@NonNull String key, boolean defValue) {
        return wrappedPreferences.getBoolean(key, defValue);
    }

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    public boolean contains(@NonNull String key) {
        return wrappedPreferences.contains(key);
    }

    @NonNull
    public List<String> getCommaList(@NonNull String key) {
        return commaSplit(getString(key, ""));
    }

    @NonNull
    public Set<String> getCommaSet(@NonNull String key) {
        return new LinkedHashSet<>(getCommaList(key));
    }

    @NonNull
    public EditorWrapper edit() {
        return new EditorWrapper(wrappedPreferences.edit());
    }

    public void registerOnSharedPreferenceChangeListener(@NonNull OnSharedPreferenceChangeListener listener) {
        wrappedPreferences.registerOnSharedPreferenceChangeListener(listener);
    }

    public void unregisterOnSharedPreferenceChangeListener(@NonNull OnSharedPreferenceChangeListener listener) {
        wrappedPreferences.unregisterOnSharedPreferenceChangeListener(listener);
    }

    public class EditorWrapper implements Editor {

        private final Editor wrappedEditor;

        protected EditorWrapper(Editor wrappedEditor) {
            this.wrappedEditor = wrappedEditor;
        }

        @Override
        public EditorWrapper putString(String key, @Nullable String value) {
            wrappedEditor.putString(key, value);
            return this;
        }

        @Override
        public EditorWrapper putStringSet(String key, @Nullable Set<String> values) {
            wrappedEditor.putStringSet(key, values);
            return this;
        }

        public void removeFromStringSet(String key, String... values) {
            Set<String> set = getStringSet(key);
            if (!set.isEmpty()) {
                set.removeAll(asList(values));
                putStringSet(key, set);
            }
        }

        public EditorWrapper putCommaSet(String key, Collection<String> set) {
            putString(key, commaJoin(set));
            return this;
        }

        @Override
        public EditorWrapper putInt(String key, int value) {
            wrappedEditor.putInt(key, value);
            return this;
        }

        @Override
        public EditorWrapper putLong(String key, long value) {
            wrappedEditor.putLong(key, value);
            return this;
        }

        @Override
        public EditorWrapper putFloat(String key, float value) {
            wrappedEditor.putFloat(key, value);
            return this;
        }

        @Override
        public EditorWrapper putBoolean(String key, boolean value) {
            wrappedEditor.putBoolean(key, value);
            return this;
        }

        public EditorWrapper putStringOptional(String key, @Nullable String value) {
            if (!contains(key)) {
                wrappedEditor.putString(key, value);
            }
            return this;
        }

        public EditorWrapper putStringSetOptional(String key, @Nullable Set<String> values) {
            if (!contains(key)) {
                wrappedEditor.putStringSet(key, values);
            }
            return this;
        }

        public EditorWrapper putIntOptional(String key, int value) {
            if (!contains(key)) {
                wrappedEditor.putInt(key, value);
            }
            return this;
        }

        public EditorWrapper putLongOptional(String key, long value) {
            if (!contains(key)) {
                wrappedEditor.putLong(key, value);
            }
            return this;
        }

        public EditorWrapper putFloatOptional(String key, float value) {
            if (!contains(key)) {
                wrappedEditor.putFloat(key, value);
            }
            return this;
        }

        public EditorWrapper putBooleanOptional(String key, boolean value) {
            if (!contains(key)) {
                wrappedEditor.putBoolean(key, value);
            }
            return this;
        }

        @Override
        public EditorWrapper remove(String key) {
            wrappedEditor.remove(key);
            return this;
        }

        @Override
        public EditorWrapper clear() {
            wrappedEditor.clear();
            return this;
        }

        @Override
        public boolean commit() {
            return wrappedEditor.commit();
        }

        @Override
        public void apply() {
            wrappedEditor.apply();
        }
    }
}

File: app/src/main/kotlin/com/bopr/android/smailer/util/SharedPreferencesWrapper.kt
package com.bopr.android.smailer.util

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener

/* Must be subclass of SharedPreferences! Otherwise it leads to unpredictable results */
open class SharedPreferencesWrapper(private val wrappedPreferences: SharedPreferences) :
    SharedPreferences {

    override fun getAll(): MutableMap<String, *> {
        return wrappedPreferences.all
    }

    override fun getString(key: String, defValue: String?): String? {
        return wrappedPreferences.getString(key, defValue)
    }

    override fun getInt(key: String, defValue: Int): Int {
        return wrappedPreferences.getInt(key, defValue)
    }

    override fun getLong(key: String, defValue: Long): Long {
        return wrappedPreferences.getLong(key, defValue)
    }

    override fun getFloat(key: String?, defValue: Float): Float {
        return wrappedPreferences.getFloat(key, defValue)
    }

    override fun getStringSet(key: String?, defValues: MutableSet<String>?): MutableSet<String>? {
        /* Must be a copy of values set. see: https://stackoverflow.com/questions/17469583/setstring-in-android-sharedpreferences-does-not-save-on-force-close */
        return wrappedPreferences.getStringSet(key, null)?.toMutableSet() ?: defValues
    }

    override fun getBoolean(key: String, defValue: Boolean): Boolean {
        return wrappedPreferences.getBoolean(key, defValue)
    }

//    inline fun <reified T : Any> get(key: String): T? {
//        return when (T::class) {
//            String::class -> getString(key) as T
//            Boolean::class -> getBoolean(key) as T
//            Int::class -> getInt(key, 0) as T
//            Long::class -> getLong(key, 0) as T
//            Float::class -> getFloat(key, 0f) as T
//            else -> null
//        }
//    }

    fun getString(key: String): String? {
        return getString(key, null)
    }

    fun requireString(key: String, defValue: String): String {
        return requireNotNull(getString(key, defValue))
    }

    fun getBoolean(key: String): Boolean {
        return getBoolean(key, false)
    }

    fun getStringSet(key: String): MutableSet<String> {
        return getStringSet(key, mutableSetOf())!!
    }

    fun getStringList(key: String): MutableList<String> {
        return getString(key)?.let { commaSplit(it).toMutableList() } ?: mutableListOf()
    }

    override fun contains(key: String): Boolean {
        return wrappedPreferences.contains(key)
    }

    override fun edit(): EditorWrapper {
        return EditorWrapper(wrappedPreferences.edit())
    }

    fun update(action: EditorWrapper.() -> Unit) {
        val editor = edit()
        action(editor)
        editor.apply()
    }

    override fun registerOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener) {
        wrappedPreferences.registerOnSharedPreferenceChangeListener(listener)
    }

    override fun unregisterOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener) {
        wrappedPreferences.unregisterOnSharedPreferenceChangeListener(listener)
    }

    open inner class EditorWrapper(private val wrappedEditor: SharedPreferences.Editor) :
        SharedPreferences.Editor {

        override fun putString(key: String, value: String?): EditorWrapper {
            wrappedEditor.putString(key, value)
            return this
        }

        override fun putStringSet(key: String, values: Set<String>?): EditorWrapper {
            wrappedEditor.putStringSet(key, values)
            return this
        }

        fun putStringList(key: String, value: Collection<String>?): EditorWrapper {
            putString(key, value?.let { commaJoin(value) })
            return this
        }

        override fun putInt(key: String, value: Int): EditorWrapper {
            wrappedEditor.putInt(key, value)
            return this
        }

        override fun putLong(key: String, value: Long): EditorWrapper {
            wrappedEditor.putLong(key, value)
            return this
        }

        override fun putFloat(key: String, value: Float): EditorWrapper {
            wrappedEditor.putFloat(key, value)
            return this
        }

        override fun putBoolean(key: String, value: Boolean): EditorWrapper {
            wrappedEditor.putBoolean(key, value)
            return this
        }

        fun putStringIfNotExists(key: String, value: String?): EditorWrapper {
            if (!contains(key)) {
                putString(key, value)
            }
            return this
        }

        fun putStringSetIfNotExists(key: String, values: Set<String>?): EditorWrapper {
            if (!contains(key)) {
                putStringSet(key, values)
            }
            return this
        }

        fun putBooleanIfNotExists(key: String, value: Boolean): EditorWrapper {
            if (!contains(key)) {
                putBoolean(key, value)
            }
            return this
        }

        override fun remove(key: String): EditorWrapper {
            wrappedEditor.remove(key)
            return this
        }

        override fun clear(): EditorWrapper {
            wrappedEditor.clear()
            return this
        }

        override fun commit(): Boolean {
            return wrappedEditor.commit()
        }

        override fun apply() {
            wrappedEditor.apply()
        }

    }

}